pymc_extras/statespace/core/properties.py

```diff
@@ -0,0 +1,257 @@
+
+    def __iter__(self) -> Iterator[T]:
+        return iter(self.items)
+
+    def __len__(self) -> int:
+        return len(self.items)
+
+    def __str__(self) -> str:
+        return f"{self.key_field}s: {list(self._index.keys())}"
+
+    def add(self, new_item: T):
+        return type(self)([*self.items, new_item])
+
+    def merge(self, other: Self, allow_duplicates: bool = False) -> Self:
+        if not isinstance(other, type(self)):
+            raise TypeError(f"Cannot merge {type(other).__name__} with {type(self).__name__}")
+
+        overlapping = set(self.names) & set(other.names)
+        if overlapping and not allow_duplicates:
+            # This is necessary for shared states
+            warnings.warn(
+                f"Duplicate names found: {overlapping}. Merge will ONLY retain unique entities",
+                UserWarning,
+            )
+            return type(self)(
+                list(self.items) + [item for item in other.items if item.name not in overlapping]
+            )
+
+        return type(self)(list(self.items) + list(other.items))
+
+    @property
+    def names(self) -> tuple[str, ...]:
+        return tuple(self._index.keys())
+
+    def copy(self) -> Info[T]:
+        return deepcopy(self)
+
+
+@dataclass(frozen=True)
+class Parameter(Property):
+    name: str
+    shape: tuple[int, ...]
+    dims: tuple[str, ...]
+    constraints: str | None = None
+
+
+@dataclass(frozen=True)
+class ParameterInfo(Info[Parameter]):
+    def __init__(self, parameters: list[Parameter]):
+        super().__init__(items=tuple(parameters), key_field="name")
+
+
+@dataclass(frozen=True)
+class Data(Property):
+    name: str
+    shape: tuple[int, ...]
+    dims: tuple[str, ...]
+    is_exogenous: bool
+
+
+@dataclass(frozen=True)
+class DataInfo(Info[Data]):
+    def __init__(self, data: list[Data]):
+        super().__init__(items=tuple(data), key_field="name")
+
+    @property
+    def needs_exogenous_data(self) -> bool:
+        return any(d.is_exogenous for d in self.items)
+
+    @property
+    def exogenous_names(self) -> tuple[str, ...]:
+        return tuple(d.name for d in self.items if d.is_exogenous)
+
+    def __str__(self) -> str:
+        return f"data: {[d.name for d in self.items]}\nneeds exogenous data: {self.needs_exogenous_data}"
+
+
+@dataclass(frozen=True)
+class Coord(Property):
+    dimension: str
+    labels: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class CoordInfo(Info[Coord]):
+    def __init__(self, coords: list[Coord]):
+        super().__init__(items=tuple(coords), key_field="dimension")
+
+    def __str__(self) -> str:
+        base = "coordinates:"
+        for coord in self.items:
+            coord_str = str(coord)
+            indented = "\n".join("  " + line for line in coord_str.splitlines())
+            base += "\n" + indented + "\n"
+        return base
+
+    @classmethod
+    def default_coords_from_model(cls, model: PyMCStateSpace | Component) -> Self:
```

Comment:
```suggestion
    def default_coords_from_model(cls, model: StateSpaceLike) -> Self:
```

To simplify all the type checking, we can just put a protocol here that has the property we expect.